drivers/SmartThings/matter-button/src/init.lua

```diff
@@ -3,172 +3,280 @@ local log = require "log"
-      end
-    elseif button_type == BUTTON_TYPE_MAP.MULTI_PRESS then
-      if state == BUTTON_STATE_MAP.HELD_COMPLETE then
-        if multi_press == 2 then
-          event = capabilities.button.button.double({state_change = true})
-        else
-          event = capabilities.button.button(string.format("pushed_%dx", multi_press))
-        end
-      end
-    end
-    if event ~= nil then
-      device:emit_event_for_endpoint(ib.endpoint_id, event)
-    end
+local function component_to_endpoint(device, component_name)
+  local map = device:get_field(ENDPOINT_TO_COMPONENT_MAP) or {}
+  for ep, component in pairs(map) do
+    if component == component_name then return ep end
   end
 end
 
---end of helper functions
---------------------------------------------------------------------------
+local function find_child(parent, ep_id)
+  return parent:get_child_by_parent_assigned_key(string.format("%02X", ep_id))
+end
+
 local function device_init(driver, device)
-  device:subscribe()
+  if device.network_type == device_lib.NETWORK_TYPE_MATTER then
+    device:subscribe()
+    device:set_find_child(find_child)
+    device:set_endpoint_to_component_fn(endpoint_to_component)
+    device:set_component_to_endpoint_fn(component_to_endpoint)
+  end
+end
+
+-- This is called either on add for parent/child devices, or after the device profile changes for components
+local function configure_buttons(device)
+  if device.network_type ~= device_lib.NETWORK_TYPE_CHILD then
+    local MS = device:get_endpoints(clusters.Switch.ID, {feature_bitmap=clusters.Switch.types.SwitchFeature.MOMENTARY_SWITCH})
+    local MSR = device:get_endpoints(clusters.Switch.ID, {feature_bitmap=clusters.Switch.types.SwitchFeature.MOMENTARY_SWITCH_RELEASE})
+    device.log.debug(#MSR.." momentary switch release endpoints")
+    local MSL = device:get_endpoints(clusters.Switch.ID, {feature_bitmap=clusters.Switch.types.SwitchFeature.MOMENTARY_SWITCH_LONG_PRESS})
+    device.log.debug(#MSL.." momentary switch long press endpoints")
+    local MSM = device:get_endpoints(clusters.Switch.ID, {feature_bitmap=clusters.Switch.types.SwitchFeature.MOMENTARY_SWITCH_MULTI_PRESS})
+    device.log.debug(#MSM.." momentary switch multi press endpoints")
+    for _, ep in ipairs(MS) do
+      local supportedButtonValues_event = capabilities.button.supportedButtonValues({"pushed", "held"}, {visibility = {displayed = false}})
+      -- this ordering is important, as MSL & MSM devices must also support MSR
+      if contains(MSM, ep) then
+        -- ask the device to tell us its max number of presses
+        device.log.debug("sending multi press max read")
+        device:send(clusters.Switch.attributes.MultiPressMax:read(device, ep))
+        set_field_for_endpoint(device, MULTI_BUTTON, ep, true, true)
+        supportedButtonValues_event = nil -- deferred until max press handler
+      elseif contains(MSL, ep) then
+        device.log.debug("configuring for long press device")
```

Comment:
It looks like this case is missing some logic

Reply:
I said that because my 4-button device supports press/hold and isn't working, but on further review it's not clear to me if the issue is here or somewhere else

Reply:
Does it also support Multipress or just Press & Hold?
Did you try to re-add the device?

Reply:
It does not support multipress. I removed/re-added a few times and the behavior was same each time.